File: EmailService.js
/**
 * EmailService - Template-based email generation and sending
 * Handles all email-related operations with retry logic
 */
const EmailService = {
  templates: {
    daily: {
      v1: {
        subjectTemplate: '{date} 每日摘要｜心情: {mood}｜行為: {behaviorScore}｜睡眠: {sleepScore}',
        
        cssStyles: `body{font-family:-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,sans-serif;line-height:1.7;color:#34495e;max-width:680px;margin:20px auto;padding:0;background-color:#f4f7f6}.container{background-color:#fff;padding:20px 30px 30px;border-radius:12px;box-shadow:0 6px 18px rgba(0,0,0,.06)}h1{color:#2c3e50;font-size:24px;border-bottom:2px solid #e0e0e0;padding-bottom:12px;margin:0 0 20px}h2{color:#2980b9;font-size:18px;margin-top:30px;margin-bottom:15px;border-left:4px solid #3498db;padding-left:12px;font-weight:600}ul{list-style-type:none;padding-left:5px;margin:0}li{margin-bottom:10px;padding-left:20px;position:relative}li::before{content:'•';color:#3498db;font-size:20px;position:absolute;left:0;top:-3px}strong{color:#c0392b;font-weight:600}.feedback-section{margin-top:30px;background-color:#e8f4fd;padding:20px;border-radius:8px;border:1px solid #d0e0f0}.feedback-section h2{margin-top:0;color:#2c3e50;border-color:#2980b9}.feedback-section li::before{color:#2980b9}.score-section{margin-top:20px;background-color:#f8f9fa;padding:15px;border-radius:8px;border:1px solid #e9ecef}.score-section h3{color:#2c3e50;margin-top:0;font-size:16px}.score-value{font-size:24px;color:#2980b9;font-weight:700;margin:10px 0}.score-details{font-size:14px;color:#7f8c8d}`,
        
        bodyTemplate: `<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><style>{css}</style></head>
<body>
<div class="container">
<h1>每日摘要報告 - {date}</h1>
<div class="score-section">
<h3>今日評分</h3>
<div class="score-value">行為效率分數：{behaviorTotal}/100</div>
<div class="score-details">(原始分: {behaviorRaw}, 正向: {behaviorPositive}, 負向: {behaviorNegative})</div>
<div class="score-value">睡眠健康指數：{sleepTotal}/100</div>
<div class="score-details">(時長: {sleepDuration}, 品質: {sleepQuality}, 規律: {sleepRegularity})</div>
</div>
{analysisContent}
</div>
</body>
</html>`
      }
    }
  },
  
  activeVersions: {
    daily: 'v1'
  },
  
  labelConfig: {
    basePath: 'Meta-Awareness',
    daily: 'Daily',
    weekly: 'Weekly'
  },
  
  retryConfig: {
    maxAttempts: 4,
    delayMs: 3000,
    firstDelayMs: 2000
  },
  
  /**
   * Send daily report email
   * @param {Object} reportData - Report data
   * @returns {Promise<boolean>} Success status
   */
  async sendDailyReport(reportData) {
    const { date, analysis, scores, mood } = reportData;
    
    // Build subject
    const subject = this.buildSubject('daily', {
      date: date,
      mood: mood || 'N/A',
      behaviorScore: scores.behavior.total,
      sleepScore: scores.sleep.total
    });
    
    // Build HTML body
    const htmlBody = this.buildHtmlBody('daily', {
      date: date,
      analysis: analysis,
      scores: scores
    });
    
    // Send email with retry logic
    console.log(`[EmailService] Sending daily report email: ${subject}`);
    
    try {
      await this._sendEmailWithRetry(CONFIG.RECIPIENT_EMAIL, subject, htmlBody);
      
      // Apply Gmail label
      await this._applyGmailLabel(CONFIG.RECIPIENT_EMAIL, subject, 'daily');
      
      console.log('[EmailService] Email sent successfully');
      return true;
      
    } catch (error) {
      console.error('[EmailService] Failed to send email:', error);
      EventBus.emit('EMAIL_SEND_FAILED', {
        error: error.message,
        recipient: CONFIG.RECIPIENT_EMAIL,
        subject: subject
      });
      throw error;
    }
  },
  
  /**
   * Build email subject from template
   * @private
   */
  buildSubject(type, data) {
    const template = this.templates[type][this.activeVersions[type]];
    let subject = template.subjectTemplate;
    
    Object.entries(data).forEach(([key, value]) => {
      subject = subject.replace(new RegExp(`{${key}}`, 'g'), value);
    });
    
    return subject;
  },
  
  /**
   * Build HTML body from template
   * @private
   */
  buildHtmlBody(type, data) {
    const template = this.templates[type][this.activeVersions[type]];
    const { date, analysis, scores } = data;
    
    // Generate analysis content HTML
    const analysisHtml = this._renderAnalysisHtml(analysis);
    
    // Prepare template data
    const templateData = {
      css: template.cssStyles,
      date: date,
      behaviorTotal: scores.behavior.total,
      behaviorRaw: scores.behavior.details?.rawScore || 0,
      behaviorPositive: scores.behavior.details?.positive || 0,
      behaviorNegative: scores.behavior.details?.negative || 0,
      sleepTotal: scores.sleep.total,
      sleepDuration: scores.sleep.details?.duration?.score || 0,
      sleepQuality: scores.sleep.details?.quality?.score || 0,
      sleepRegularity: scores.sleep.details?.regularity?.score || 0,
      analysisContent: analysisHtml
    };
    
    // Replace placeholders
    let html = template.bodyTemplate;
    Object.entries(templateData).forEach(([key, value]) => {
      html = html.replace(new RegExp(`{${key}}`, 'g'), value);
    });
    
    return html;
  },
  
  /**
   * Render analysis JSON to HTML
   * @private
   */
  _renderAnalysisHtml(analysis) {
    if (!analysis) {
      return '<h2>分析生成失敗</h2><p>無法從 AI 服務獲取有效的分析結果。</p>';
    }
    
    const escapeHtml = (text) => {
      if (typeof text !== 'string') return '';
      return text.replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#039;');
    };
    
    const br = analysis.behaviorReview || {};
    const sa = analysis.sleepAnalysis || {};
    const ss = analysis.statusSummary || {};
    const cf = analysis.coachFeedback || {};
    
    return `
      <h2>${escapeHtml(br.title || '1. 行為盤點')}</h2>
      <ul>
        <li><strong>正向行為</strong>：${(br.positive && br.positive.length > 0) ? escapeHtml(br.positive.join('、')) : '無'}</li>
        <li><strong>負向行為</strong>：${(br.negative && br.negative.length > 0) ? escapeHtml(br.negative.join('、')) : '無'}</li>
        <li><strong>中性行為</strong>：${(br.neutral && br.neutral.length > 0) ? escapeHtml(br.neutral.join('、')) : '無'}</li>
      </ul>
      
      <h2>${escapeHtml(sa.title || '2. 睡眠分析與洞察')}</h2>
      <ul>
        ${sa.insight ? `<li>${escapeHtml(sa.insight)}</li>` : ''}
        ${(sa.recommendations || []).map(rec => `<li>${escapeHtml(rec)}</li>`).join('')}
      </ul>
      
      <h2>${escapeHtml(ss.title || '3. 今日狀態短評')}</h2>
      <ul>
        ${ss.comment ? `<li>${escapeHtml(ss.comment)}</li>` : ''}
        ${ss.highlights && ss.highlights.length > 0 ? `<li><strong>亮點</strong>：${escapeHtml(ss.highlights.join('、'))}</li>` : ''}
      </ul>

      <div class="feedback-section">
        <h2>${escapeHtml(cf.title || '4. AI 教練回饋與提問')}</h2>
        <ul>
          ${cf.affirmation ? `<li><strong>肯定</strong>：${escapeHtml(cf.affirmation)}</li>` : ''}
          ${cf.suggestion ? `<li><strong>建議</strong>：${escapeHtml(cf.suggestion)}</li>` : ''}
          ${cf.opportunity ? `<li><strong>機會點</strong>：${escapeHtml(cf.opportunity)}</li>` : ''}
          ${cf.question ? `<li><strong>問題</strong>：${escapeHtml(cf.question)}</li>` : ''}
        </ul>
      </div>
    `;
  },
  
  /**
   * Send email with retry logic
   * @private
   */
  async _sendEmailWithRetry(recipient, subject, htmlBody) {
    const maxAttempts = 1; // Email sending usually doesn't need retry
    
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      try {
        GmailApp.sendEmail(recipient, subject, "", { htmlBody: htmlBody });
        return;
      } catch (error) {
        if (attempt === maxAttempts) {
          throw error;
        }
        console.log(`[EmailService] Email send failed, attempt ${attempt}/${maxAttempts}`);
        Utilities.sleep(this.retryConfig.delayMs);
      }
    }
  },
  
  /**
   * Apply Gmail label to sent email
   * @private
   */
  async _applyGmailLabel(recipient, subject, reportType) {
    const labelPath = `${this.labelConfig.basePath}/${reportType === 'daily' ? this.labelConfig.daily : this.labelConfig.weekly}`;
    const label = GmailApp.getUserLabelByName(labelPath) || GmailApp.createLabel(labelPath);
    
    // Sanitize subject for search
    const sanitizedSubject = subject.replace(/[|]/g, ' ').replace(/\s+/g, ' ').trim();
    const searchQuery = `subject:("${sanitizedSubject}") to:(${recipient}) from:me newer_than:2m`;
    
    let attempts = 0;
    let delay = this.retryConfig.firstDelayMs;
    
    while (attempts < this.retryConfig.maxAttempts) {
      attempts++;
      
      if (attempts > 1) {
        Utilities.sleep(delay);
        delay = this.retryConfig.delayMs;
      } else {
        Utilities.sleep(this.retryConfig.firstDelayMs);
      }
      
      console.log(`[EmailService] Searching for email to label (attempt ${attempts}/${this.retryConfig.maxAttempts})`);
      const threads = GmailApp.search(searchQuery, 0, 1);
      
      if (threads.length > 0) {
        threads[0].addLabel(label);
        console.log(`[EmailService] Successfully applied label '${labelPath}'`);
        return;
      }
    }
    
    console.warn('[EmailService] Could not find email to apply label');
  },
  
  /**
   * Register email template
   * @param {string} type - Template type
   * @param {string} version - Version identifier
   * @param {Object} template - Template definition
   */
  registerTemplate(type, version, template) {
    if (!this.templates[type]) {
      this.templates[type] = {};
    }
    
    this.templates[type][version] = template;
    console.log(`[EmailService] Registered ${type} template version ${version}`);
  }
}; 

File: src/email_service.py
"""Email service for sending diary analysis results"""

import smtplib
import ssl
import time
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.utils import formataddr
from typing import Optional, Dict, Any
from datetime import datetime
import json

try:
    from googleapiclient.discovery import build
    from google.oauth2.service_account import Credentials
    GMAIL_API_AVAILABLE = True
except ImportError:
    GMAIL_API_AVAILABLE = False

from .models import InsightPack
from .config import Config
from .logger import get_logger

logger = get_logger(__name__)


class EmailService:
    """Handles sending analysis results via email using SMTP"""

    def __init__(self, config: Config):
        self.config = config
        self.gmail_service = None

    def send_analysis_result(self, insight_pack: InsightPack, run_id: str) -> bool:
        """
        Send diary analysis results via email
        
        Args:
            insight_pack: Analysis results to send
            run_id: Run identifier for tracking
            
        Returns:
            bool: Success status
        """
        if not self._validate_email_config():
            logger.warning("Email not configured, skipping email send")
            return False

        try:
            subject = self._build_subject(insight_pack)
            html_body = self._build_html_body(insight_pack)
            
            success = self._send_email(
                recipient=self.config.EMAIL_RECIPIENT,
                subject=subject,
                html_body=html_body,
                run_id=run_id
            )
            
            if success:
                logger.info("Analysis email sent successfully to %s", self.config.EMAIL_RECIPIENT)
                
                # Apply Gmail label if enabled
                if getattr(self.config, 'EMAIL_APPLY_LABEL', True):
                    try:
                        self._apply_gmail_label(subject, self.config.EMAIL_RECIPIENT)
                    except Exception as e:
                        logger.warning("Gmail labeling failed (email still sent): %s", str(e))
                        logger.info("💡 Tip: Create Gmail filter - From: %s, Subject: '日誌分析報告', Label: 'Meta-Awareness/Weekly'", 
                                  self.config.EMAIL_SENDER)
            else:
                logger.error("Failed to send analysis email")
                
            return success
            
        except Exception as e:
            logger.error("Email sending failed: %s", str(e))
            return False

    def _validate_email_config(self) -> bool:
        """Validate email configuration is complete"""
        required_fields = [
            'EMAIL_SMTP_SERVER',
            'EMAIL_SMTP_PORT', 
            'EMAIL_SENDER',
            'EMAIL_PASSWORD',
            'EMAIL_RECIPIENT'
        ]
        
        missing_fields = []
        for field in required_fields:
            value = getattr(self.config, field, None)
            if not value:
                missing_fields.append(field)
                logger.debug("Missing email config field: %s (value: %s)", field, value)
        
        if missing_fields:
            logger.warning("Email config incomplete. Missing fields: %s", missing_fields)
            return False
            
        logger.debug("Email config validation passed")
        return True

    def _send_email(self, recipient: str, subject: str, html_body: str, run_id: str) -> bool:
        """Send email via SMTP with retry logic"""
        max_retries = getattr(self.config, 'EMAIL_MAX_RETRIES', 2)
        
        for attempt in range(max_retries):
            try:
                # Create message
                message = MIMEMultipart("alternative")
                message["Subject"] = subject
                message["From"] = formataddr((
                    getattr(self.config, 'EMAIL_SENDER_NAME', '三省日誌分析'),
                    self.config.EMAIL_SENDER
                ))
                message["To"] = recipient
                message["X-Run-ID"] = run_id

                # Add HTML content
                html_part = MIMEText(html_body, "html", "utf-8")
                message.attach(html_part)

                # Send via SMTP
                context = ssl.create_default_context()
                
                with smtplib.SMTP(self.config.EMAIL_SMTP_SERVER, self.config.EMAIL_SMTP_PORT) as server:
                    server.starttls(context=context)
                    server.login(self.config.EMAIL_SENDER, self.config.EMAIL_PASSWORD)
                    server.send_message(message)
                
                logger.info("Email sent successfully on attempt %d", attempt + 1)
                return True
                
            except Exception as e:
                logger.warning("Email attempt %d failed: %s", attempt + 1, str(e))
                if attempt == max_retries - 1:
                    logger.error("All email attempts failed")
                    return False
        
        return False

    def _build_subject(self, insight_pack: InsightPack) -> str:
        """Build email subject from analysis data"""
        date_str = datetime.now().strftime("%Y-%m-%d")
        entry_count = insight_pack.meta.get("entriesAnalyzed", 0)
        
        # Try to get primary theme if available
        primary_theme = ""
        if insight_pack.themes:
            primary_theme = f" | {insight_pack.themes[0].label}"
        
        return f"日誌分析報告 {date_str} | {entry_count} 篇{primary_theme}"

    def _build_html_body(self, insight_pack: InsightPack) -> str:
        """Build HTML email body from analysis results"""
        # Generate HTML content
        analysis_html = self._render_analysis_html(insight_pack)
        
        # CSS styles for email
        css_styles = """
        <style>
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, sans-serif;
            line-height: 1.6;
            color: #34495e;
            max-width: 680px;
            margin: 20px auto;
            padding: 0;
            background-color: #f8f9fa;
        }
        .container {
            background-color: #fff;
            padding: 30px;
            border-radius: 12px;
            box-shadow: 0 4px 12px rgba(0,0,0,0.1);
        }
        .header {
            text-align: center;
            padding-bottom: 20px;
            border-bottom: 2px solid #e9ecef;
            margin-bottom: 30px;
        }
        .header h1 {
            color: #2c3e50;
            margin: 0;
            font-size: 24px;
        }
        .meta-info {
            color: #6c757d;
            font-size: 14px;
            margin-top: 10px;
        }
        .section {
            margin-bottom: 30px;
        }
        .section-title {
            color: #2980b9;
            font-size: 18px;
            margin-bottom: 15px;
            padding-left: 12px;
            border-left: 4px solid #3498db;
            font-weight: 600;
        }
        .daily-summaries {
            background-color: #f8f9fa;
            padding: 20px;
            border-radius: 8px;
            margin-bottom: 20px;
        }
        .daily-item {
            margin-bottom: 12px;
            padding: 10px;
            background-color: #fff;
            border-radius: 6px;
            border-left: 3px solid #3498db;
        }
        .daily-date {
            font-weight: 600;
            color: #2c3e50;
            font-size: 14px;
        }
        .daily-summary {
            margin-top: 5px;
            color: #495057;
        }
        .themes-grid {
            display: grid;
            grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
            gap: 15px;
            margin-top: 15px;
        }
        .theme-item {
            background-color: #e8f4f8;
            padding: 15px;
            border-radius: 8px;
            text-align: center;
            border: 1px solid #bee5eb;
        }
        .theme-label {
            font-weight: 600;
            color: #0056b3;
            font-size: 16px;
            margin-bottom: 5px;
        }
        .theme-support {
            color: #6c757d;
            font-size: 14px;
        }
        .reflection-box {
            background-color: #fff3cd;
            border: 1px solid #ffeaa7;
            border-radius: 8px;
            padding: 20px;
            margin-top: 15px;
        }
        .reflection-question {
            color: #856404;
            font-size: 16px;
            font-weight: 500;
            font-style: italic;
            text-align: center;
        }
        .insights-list {
            list-style: none;
            padding: 0;
        }
        .insights-list li {
            background-color: #f8f9fa;
            margin-bottom: 10px;
            padding: 12px 15px;
            border-radius: 6px;
            border-left: 3px solid #28a745;
            position: relative;
        }
        .insights-list li::before {
            content: '•';
            color: #28a745;
            margin-right: 10px;
        }
        .footer {
            margin-top: 40px;
            padding-top: 20px;
            border-top: 1px solid #e9ecef;
            text-align: center;
            color: #6c757d;
            font-size: 14px;
        }
        .run-info {
            background-color: #f1f3f4;
            padding: 10px 15px;
            border-radius: 6px;
            margin-top: 15px;
            font-size: 12px;
            color: #5f6368;
        }
        </style>
        """
        
        return f"""
        <!DOCTYPE html>
        <html lang="zh-TW">
        <head>
            <meta charset="UTF-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title>日誌分析報告</title>
            {css_styles}
        </head>
        <body>
            <div class="container">
                {analysis_html}
            </div>
        </body>
        </html>
        """

    def _render_analysis_html(self, insight_pack: InsightPack) -> str:
        """Render InsightPack to HTML"""
        # Header section
        run_id = insight_pack.meta.get("run_id", "unknown")
        entry_count = insight_pack.meta.get("entriesAnalyzed", 0)
        generated_at = insight_pack.meta.get("generatedAt", "")
        
        html_parts = [
            '<div class="header">',
            '<h1>日誌分析報告</h1>',
            f'<div class="meta-info">分析了 {entry_count} 篇日誌 · {generated_at[:10] if generated_at else ""}</div>',
            '</div>'
        ]
        
        # Daily summaries section
        if insight_pack.dailySummaries:
            html_parts.extend([
                '<div class="section">',
                '<h2 class="section-title">每日摘要</h2>',
                '<div class="daily-summaries">'
            ])
            
            for summary in insight_pack.dailySummaries:
                html_parts.extend([
                    '<div class="daily-item">',
                    f'<div class="daily-date">{summary.date}</div>',
                    f'<div class="daily-summary">{self._escape_html(summary.summary)}</div>',
                    '</div>'
                ])
            
            html_parts.extend(['</div>', '</div>'])
        
        # Themes section
        if insight_pack.themes:
            html_parts.extend([
                '<div class="section">',
                '<h2 class="section-title">核心主題</h2>',
                '<div class="themes-grid">'
            ])
            
            for theme in insight_pack.themes:
                html_parts.extend([
                    '<div class="theme-item">',
                    f'<div class="theme-label">{self._escape_html(theme.label)}</div>',
                    f'<div class="theme-support">支持度: {theme.support}</div>',
                    '</div>'
                ])
            
            html_parts.extend(['</div>', '</div>'])
        
        # Hidden signals section
        if insight_pack.hiddenSignals:
            html_parts.extend([
                '<div class="section">',
                '<h2 class="section-title">潛在信號</h2>',
                '<ul class="insights-list">'
            ])
            
            for signal in insight_pack.hiddenSignals:
                html_parts.append(f'<li>{self._escape_html(signal)}</li>')
            
            html_parts.extend(['</ul>', '</div>'])
        
        # Emotional indicators section
        if insight_pack.emotionalIndicators:
            html_parts.extend([
                '<div class="section">',
                '<h2 class="section-title">情緒指標</h2>',
                '<ul class="insights-list">'
            ])
            
            for indicator in insight_pack.emotionalIndicators:
                html_parts.append(f'<li>{self._escape_html(str(indicator))}</li>')
            
            html_parts.extend(['</ul>', '</div>'])
        
        # Anomalies section
        if insight_pack.anomalies:
            html_parts.extend([
                '<div class="section">',
                '<h2 class="section-title">異常觀察</h2>',
                '<ul class="insights-list">'
            ])
            
            for anomaly in insight_pack.anomalies:
                html_parts.append(f'<li>{self._escape_html(anomaly)}</li>')
            
            html_parts.extend(['</ul>', '</div>'])
        
        # Reflection question section
        if insight_pack.reflectiveQuestion:
            html_parts.extend([
                '<div class="section">',
                '<h2 class="section-title">反思問題</h2>',
                '<div class="reflection-box">',
                f'<div class="reflection-question">{self._escape_html(insight_pack.reflectiveQuestion)}</div>',
                '</div>',
                '</div>'
            ])
        
        # Footer
        html_parts.extend([
            '<div class="footer">',
            '<div>三省日誌分析引擎 · 用心記錄，智慧成長</div>',
            f'<div class="run-info">Run ID: {run_id}</div>',
            '</div>'
        ])
        
        return ''.join(html_parts)

    def _escape_html(self, text: str) -> str:
        """Escape HTML special characters"""
        if not isinstance(text, str):
            text = str(text)
        return (text.replace('&', '&amp;')
                   .replace('<', '&lt;')
                   .replace('>', '&gt;')
                   .replace('"', '&quot;')
                   .replace("'", '&#x27;'))

    def _init_gmail_service(self):
        """Initialize Gmail API service using service account credentials"""
        if not GMAIL_API_AVAILABLE:
            logger.debug("Gmail API not available (missing google-api-python-client)")
            return None
            
        if not hasattr(self.config, 'CREDENTIALS_PATH') or not self.config.CREDENTIALS_PATH:
            logger.debug("No credentials path for Gmail API")
            return None
            
        try:
            # Use the same service account as the Sheets API
            credentials = Credentials.from_service_account_file(
                str(self.config.CREDENTIALS_PATH),
                scopes=['https://www.googleapis.com/auth/gmail.modify']
            )
            
            # Delegate to the user's email (impersonate)
            if hasattr(credentials, 'with_subject'):
                delegated_credentials = credentials.with_subject(self.config.EMAIL_SENDER)
                service = build('gmail', 'v1', credentials=delegated_credentials)
            else:
                logger.warning("Cannot delegate Gmail API access - labels will not work")
                return None
                
            logger.debug("Gmail API service initialized successfully")
            return service
            
        except Exception as e:
            logger.debug("Failed to initialize Gmail API: %s", str(e))
            return None

    def _apply_gmail_label(self, subject: str, recipient: str):
        """Apply Gmail label to sent email"""
        if not self.gmail_service:
            self.gmail_service = self._init_gmail_service()
            
        if not self.gmail_service:
            logger.debug("Gmail labeling skipped - API not available")
            return
            
        try:
            # Wait a moment for email to be indexed
            time.sleep(3)
            
            # Search for the sent email
            query = f'to:{recipient} subject:"{subject}" from:me newer_than:5m'
            logger.debug("Searching Gmail with query: %s", query)
            
            search_result = self.gmail_service.users().messages().list(
                userId='me',
                q=query,
                maxResults=1
            ).execute()
            
            messages = search_result.get('messages', [])
            if not messages:
                logger.warning("Could not find sent email for labeling")
                return
                
            message_id = messages[0]['id']
            
            # Get or create the label
            label_name = getattr(self.config, 'EMAIL_GMAIL_LABEL', 'Meta-Awareness/Weekly')
            label_id = self._get_or_create_label(label_name)
            
            if label_id:
                # Apply the label
                self.gmail_service.users().messages().modify(
                    userId='me',
                    id=message_id,
                    body={'addLabelIds': [label_id]}
                ).execute()
                
                logger.info("Gmail label '%s' applied successfully", label_name)
            else:
                logger.warning("Could not create Gmail label")
                
        except Exception as e:
            logger.warning("Failed to apply Gmail label: %s", str(e))

    def _get_or_create_label(self, label_name: str) -> Optional[str]:
        """Get existing label or create new one"""
        try:
            # List existing labels
            labels_result = self.gmail_service.users().labels().list(userId='me').execute()
            labels = labels_result.get('labels', [])
            
            # Check if label already exists
            for label in labels:
                if label['name'] == label_name:
                    logger.debug("Found existing Gmail label: %s", label_name)
                    return label['id']
            
            # Create new label
            logger.debug("Creating new Gmail label: %s", label_name)
            label_body = {
                'name': label_name,
                'labelListVisibility': 'labelShow',
                'messageListVisibility': 'show'
            }
            
            created_label = self.gmail_service.users().labels().create(
                userId='me',
                body=label_body
            ).execute()
            
            logger.info("Created Gmail label: %s", label_name)
            return created_label['id']
            
        except Exception as e:
            logger.warning("Failed to get/create Gmail label: %s", str(e))
            return None